plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,698 @@
+            this.sseCustomerKey = requireNonNull(sseCustomerKey, "sseCustomerKey is null");
+        }
+
+        @Override
+        public void writeInt(int value)
+        {
+            writeBytes(ByteBuffer.allocate(Integer.BYTES).order(ByteOrder.LITTLE_ENDIAN).putInt(value).array());
+        }
+
+        @Override
+        public void writeBytes(Slice slice)
+        {
+            writeBytes(slice.getBytes());
+        }
+
+        @Override
+        public void writeBytes(byte[] buffer)
+        {
+            writeBytes(buffer, 0, buffer.length);
+        }
+
+        @Override
+        public void writeBytes(byte[] bytes, int offset, int length)
+        {
+            while (length > 0) {
+                int copied = min(buffer.length - bufferSize, length);
+                arraycopy(bytes, offset, buffer, bufferSize, copied);
+                bufferSize += copied;
+
+                try {
+                    flushBuffer(false);
+                }
+                catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+
+                offset += copied;
+                length -= copied;
+            }
+        }
+
+        @Override
+        public long getRetainedSize()
+        {
+            return sizeOf(buffer) + INSTANCE_SIZE;
+        }
+
+        @Override
+        public void close()
+                throws IOException
+        {
+            if (failed) {
+                try {
+                    abortUpload();
+                    return;
+                }
+                catch (RuntimeException e) {
+                    throw new IOException(e);
+                }
+            }
+
+            try {
+                flushBuffer(true);
+                waitForPreviousUploadFinish();
+            }
+            catch (IOException | RuntimeException e) {
+                abortUploadSuppressed(e);
+                throw e;
+            }
+
+            try {
+                uploadId.ifPresent(this::finishUpload);
+            }
+            catch (RuntimeException e) {
+                abortUploadSuppressed(e);
+                throw new IOException(e);
+            }
+        }
+
+        private void flushBuffer(boolean finished)
+                throws IOException
+        {
+            // skip multipart upload if there would only be one part
+            if (finished && uploadId.isEmpty()) {
+                InputStream in = new ByteArrayInputStream(buffer, 0, bufferSize);
+
+                ObjectMetadata metadata = new ObjectMetadata();
+                metadata.setContentLength(bufferSize);
+                metadata.setContentMD5(getMd5AsBase64(buffer, 0, bufferSize));
+
+                PutObjectRequest request = new PutObjectRequest(bucketName, key, in, metadata);
+                sseCustomerKey.ifPresent(request::withSSECustomerKey);
+
+                try {
+                    s3.putObject(request);
```

Comment:
Does S3 client offer a non blocking (async) method?

Reply:
Yes, however I need some time researching it on how much benefit it will bring to us.